drivers/dma/dma_iproc_pax.h

```diff
@@ -0,0 +1,524 @@
+					 PAX_DMA_STATUS_PCIE_RX_POISON \
+					)
+/* completion RM status code */
+#define RM_COMPLETION_SUCCESS		0x0
+#define RM_COMPLETION_AE_TIMEOUT	0x3FF
+
+#define RM_COMM_MSI_CONFIG_INTERRUPT_ACCESS_ERR_MASK	BIT(9)
+#define RM_COMM_MSI_CONFIG_INTERRUPT_BRESP_ERR_MASK	BIT(8)
+#define RM_COMM_MSI_DISABLE_MASK			BIT(0)
+
+/* Buffer Descriptor definitions */
+#define PAX_DMA_TYPE_RM_HEADER		0x1
+#define PAX_DMA_TYPE_NEXT_PTR		0x5
+
+/* one desc ring size( is 4K, 4K aligned */
+#define PAX_DMA_RM_DESC_RING_SIZE	4096
+#define PAX_DMA_RING_BD_ALIGN_ORDER	12
+/* completion ring size(bytes) is 8K, 8K aligned */
+#define PAX_DMA_RM_CMPL_RING_SIZE	8192
+#define PAX_DMA_RING_CMPL_ALIGN_ORDER	13
+
+#define PAX_DMA_RING_BD_ALIGN_CHECK(addr) \
+	(!((addr) & ((0x1 << RING_BD_ALIGN_ORDER) - 1)))
+#define RING_CMPL_ALIGN_CHECK(addr) \
+	(!((addr) & ((0x1 << RING_CMPL_ALIGN_ORDER) - 1)))
+
+/* RM descriptor width: 8 bytes */
+#define PAX_DMA_RM_DESC_BDWIDTH		8
+/* completion msg desc takes 1 BD */
+#define PAX_DMA_CMPL_DESC_SIZE		PAX_DMA_RM_DESC_BDWIDTH
+/* Next table desc takes 1 BD */
+#define PAX_DMA_NEXT_TBL_DESC_SIZE	PAX_DMA_RM_DESC_BDWIDTH
+/* Header desc takes 1 BD */
+#define PAX_DMA_HEADER_DESC_SIZE	PAX_DMA_RM_DESC_BDWIDTH
+/* Total BDs in ring: 4K/8bytes = 512 BDs */
+#define PAX_DMA_RM_RING_BD_COUNT	(PAX_DMA_RM_DESC_RING_SIZE / \
+					 PAX_DMA_RM_DESC_BDWIDTH)
+
+/* Initial RM header is first BD in ring */
+#define PAX_DMA_HEADER_INDEX		0
+#define PAX_DMA_HEADER_ADDR(_ring)	(void *)((uintptr_t)(_ring) + \
+	PAX_DMA_HEADER_INDEX * PAX_DMA_RM_DESC_BDWIDTH)
+
+/* NEXT TABLE desc offset is last BD in ring */
+#define PAX_DMA_NEXT_TBL_INDEX		(PAX_DMA_RM_RING_BD_COUNT - 1)
+#define PAX_DMA_NEXT_TBL_ADDR(_ring)	(void *)((uintptr_t)(_ring) + \
+	PAX_DMA_NEXT_TBL_INDEX * PAX_DMA_RM_DESC_BDWIDTH)
+
+/* DMA transfers supported from 4 bytes thru 16M, size aligned to 4 bytes */
+#define PAX_DMA_MIN_SIZE	4
+#define PAX_DMA_MAX_SIZE	(16 * 1024 * 1024)
+
+/* Host and Card address need 4-byte alignment */
+#define PAX_DMA_ADDR_ALIGN 4
+
+#define RM_RING_REG(_pd, _r, _write_ptr) \
+	((_pd)->ring[_r].ring_base + (_write_ptr))
+#define RM_COMM_REG(_pd, _write_ptr) ((_pd)->rm_comm_base + (_write_ptr))
+#define PAX_DMA_REG(_pd, _write_ptr) ((_pd)->dma_base + (_write_ptr))
+
+#define PAX_DMA_MAX_CMPL_COUNT		1024
+#define PAX_DMA_LAST_CMPL_IDX		(PAX_DMA_MAX_CMPL_COUNT - 1)
+
+#define PAX_DMA_RING_ALIGN		BIT(PAX_DMA_RING_CMPL_ALIGN_ORDER)
+/* num of completions received, circular buffer */
+#define PAX_DMA_GET_CMPL_COUNT(wptr, rptr) (((wptr) >= (rptr)) ? \
+	((wptr) - (rptr)) : (PAX_DMA_MAX_CMPL_COUNT - (rptr) + (wptr)))
+
+/* location of current cmpl pkt, take care of pointer wrap-around */
+#define PAX_DMA_CURR_CMPL_IDX(wptr)	\
+	(((wptr) == 0) ? PAX_DMA_LAST_CMPL_IDX : (wptr) - 1)
+
+/* Timeout (milliseconds) for completion alert in interrupt mode */
+#define PAX_DMA_TIMEOUT			10000
+
+/* TODO: add macro to enable memory barrier */
+#define dma_mb()
```

Comment:
As mentioned in the use location of this, more clarity is needed here -- what's the intent of this macro, and what are the semantics? (even if it's not implemented right now).

Reply:
here dma_mb() should implement a macro to enable data memory barrier, to ensure writes to memory happen.